pkg/tools/cli/loadtest/usecase/load.go

```diff
@@ -18,27 +18,99 @@ package usecase
 
 import (
 	"context"
+	"fmt"
 
 	"github.com/vdaas/vald/internal/errgroup"
+	"github.com/vdaas/vald/internal/errors"
+	"github.com/vdaas/vald/internal/log"
+	"github.com/vdaas/vald/internal/net/grpc"
 	"github.com/vdaas/vald/internal/runner"
-	"github.com/vdaas/vald/pkg/agent/ngt/config"
+	"github.com/vdaas/vald/internal/safety"
+	"github.com/vdaas/vald/pkg/tools/cli/loadtest/config"
+	"github.com/vdaas/vald/pkg/tools/cli/loadtest/service"
+	"github.com/vdaas/vald/pkg/tools/cli/loadtest/service/insert"
+	"github.com/vdaas/vald/pkg/tools/cli/loadtest/service/search"
 )
 
 type run struct {
-	eg  errgroup.Group
-	cfg *config.Data
+	eg     errgroup.Group
+	cfg    *config.Data
+	loader service.Load
+	client grpc.Client
 }
 
 func New(cfg *config.Data) (r runner.Runner, err error) {
-	return nil, nil
+	run := &run{
+		cfg: cfg,
+		eg:  errgroup.Get(),
+	}
+
+	return run, nil
 }
 
-func (r *run) PreStart(ctx context.Context) error {
-	return nil
+func (r *run) PreStart(ctx context.Context) (err error) {
+	r.client = grpc.New(
+		grpc.WithAddrs(append([]string{r.cfg.Addr}, r.cfg.Client.Addrs...)...),
+		grpc.WithInsecure(r.cfg.Client.DialOption.Insecure),
+		grpc.WithErrGroup(r.eg),
+	)
+	switch Atoo(r.cfg.Method) {
+	case Insert:
+		r.loader, err = insert.New(insert.WithAddr(r.cfg.Addr), insert.WithDataset(r.cfg.Dataset), insert.WithClient(r.client))
+	case Search:
+		r.loader, err = search.New(search.WithAddr(r.cfg.Addr), search.WithDataset(r.cfg.Dataset), search.WithClient(r.client))
+	default:
+		return fmt.Errorf("unsupported method")
+	}
+
+	return r.loader.Prepare(ctx)
 }
 
 func (r *run) Start(ctx context.Context) (<-chan error, error) {
-	return nil, nil
+	rech, err := r.client.StartConnectionMonitor(ctx)
+	if err != nil {
+		return nil, err
+	}
+	lech := r.loader.Do(ctx)
+	ech := make(chan error, 1000)
+	r.eg.Go(safety.RecoverFunc(func() (err error) {
+		defer close(ech)
+		finalize := func() (err error) {
+			var errs error
+			if r.client != nil {
+				err = r.client.Close()
+				if err != nil {
+					errs = errors.Wrap(errs, err.Error())
+				}
+			}
+			err = ctx.Err()
+			if err != nil && err != context.Canceled {
+				errs = errors.Wrap(errs, err.Error())
+			}
+			return errs
+		}
+		for {
+			select {
+			case <-ctx.Done():
+				return finalize()
+			case err = <-rech:
+			case err = <-lech:
+				if err != nil {
+					ech <- err
+					err = nil
+				}
+			}
+			if err != nil {
```

Comment:
if think this checking will cost 100% cpu usage, can you confirm it?
if yes please consider moving this block to outside of the for loop

Reply:
Passing here is only getting error from rech or lech, so I think this loop is not busy.
I checked this with following code sample.
However, I think a program like this might be a little redundant.
```
package main

import (
	"context"
	"fmt"
	"time"
)

func main() {
	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()

	ech := make(chan error, 10)
	defer close(ech)
	aech := make(chan error, 10)
	defer close(aech)
	
	go func() {
		var err error
		for {
			select {
			case <-ctx.Done():
				return
			case err = <-aech:
				//fmt.Printf("recieve error from aech: %v\n", err)
			}
			if err != nil {
				select {
				case <-ctx.Done():
					return
				case ech <- err:
					fmt.Println("send error to ech")
				}
			}
		}
	}()

	go func() {
		var err error
		for {
			select {
			case <-ctx.Done():
				return
			case err = <-ech:
				//fmt.Printf("recieve error from ech: %v\n", err)
				_ = err
			}
		}
	}()

	go func() {
		for {
			select {
			case <-ctx.Done():
				return
			case aech <- fmt.Errorf("error"):
				time.Sleep(1 * time.Second)
			}
		}
	}()
	
	time.Sleep(10 * time.Second)
}
```

```
$ go run main.go
send error to ech
send error to ech
send error to ech
send error to ech
send error to ech
send error to ech
send error to ech
send error to ech
send error to ech
send error to ech
```

Reply:
thank you. I forgot the `select` statement is a blocking statement.